### DISCUSSION TITLE
Adding New Lines Instead of Updating Existing Ones for Minor Version Upgrades

### DISCUSSION THREAD

**Discussion by ned-si** [renovatebot/renovate#30843] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

Hey there,

In our `versions.yaml` file for Kubernetes, we need to maintain multiple minor versions simultaneously while ensuring that patch releases for those versions are updated in place. However, when a new minor version is released, we require that it be added as a new entry rather than updating an existing one.

Currently, Renovate is configured to update the version number in place, but for our use case, we require that new minor versions are appended as new lines with the associated comment, instead of modifying the existing version lines. 

This is our current version file:

```yaml
versions:
# renovate: datasource=github-releases depName=kubernetes/kubernetes
- "1.27.16"
# renovate: datasource=github-releases depName=kubernetes/kubernetes
- "1.28.12"
# renovate: datasource=github-releases depName=kubernetes/kubernetes
- "1.29.8"
# renovate: datasource=github-releases depName=kubernetes/kubernetes
- "1.30.4"
# renovate: datasource=github-releases depName=kubernetes/kubernetes
- "1.31.0"
```

When a new minor version (e.g., `1.32.0`) is released, Renovate should add a new line like this:

```yaml
# renovate: datasource=github-releases depName=kubernetes/kubernetes
- "1.32.0"
```

Currently, we manage to make it work for patch releases with the following config:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "configMigration": true,
  "extends": [
    ":dependencyDashboard",
    ":semanticPrefixFixDepsChoreOthers",
    ":ignoreModulesAndTests",
    "group:monorepos",
    "group:recommended",
    "replacements:all",
    "workarounds:all"
  ],
  "customManagers": [
    {
      "customType": "regex",
      "description": "Update kubernetes versions",
      "fileMatch": [
        ".*"
      ],
      "matchStrings": [
        "renovate: datasource=(?<datasource>[a-z-]+)(?: depName=(?<depName>[^\\s]+))?.*\\n.*(?<currentValue>\\d+\\.\\d+\\.\\d+)"
      ]
    }
  ],
  "ignorePaths": [
    "manifests/**",
    "tests/**"
  ],
  "packageRules": [
    {
      "matchDatasources": [
        "github-releases"
      ],
      "matchDepNames": [
        "kubernetes/kubernetes"
      ],
      "matchPackageNames": [
        "kubernetes/kubernetes"
      ],
      "matchUpdateTypes": [
        "major",
        "minor"
      ],
      "enabled": false
    },
    {
      "matchDatasources": [
        "github-releases"
      ],
      "matchDepNames": [
        "kubernetes/kubernetes"
      ],
      "matchPackageNames": [
        "kubernetes/kubernetes"
      ],
      "separateMinorPatch": true,
      "schedule": [
        "after 8pm every weekday",
        "every weekend",
        "before 5am every weekday"
      ]
    }
  ],
  "schedule": [
    "every weekend"
  ]
}
```

(I'm not sure the `separateMinorPatch` is really needed tho)

The problem with our current configuration is that when a new minor release arrives, it's not automatically added, and we have to insert the new line manually. Ideally, we'd like to retain the existing behavior for patch releases—updating the current version in place—while automatically adding a new line whenever a new minor release occurs.

Would it be possible to configure Renovate to handle this scenario, allowing us to maintain the current behavior for existing lines and append a new entry for new minor versions?

It seems something like `autoReplaceStringTemplate` might help here, but I haven't figured out a way to make it work without breaking the existing behavior.

Thanks in advance for your help!

Cheers

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by RahulGautamSingh** [comment, marked as answer]
Not possible currently. 

You will need 2 custom regex managers for this. One will directly update the version strings(say `patchMgr`) and the other will use the comment to do so, which will also add the new line using `autoReplaceStringTemplate` (say `minorMgr`). Then we will need to block `patch` updates from `minorMgr` using `packageRules` and viceversa.

But, we do not have the capability to name the custom managers, yet.

There's an open issue for it: https://github.com/renovatebot/renovate/issues/21760.

**Comment by ned-si** [comment]
Thanks for your quick answer!

> But, we do not have the capability to name the custom managers, yet.

It seems like a pretty significant change to implement, but hopefully it'll be possible soon tho :crossed_fingers:

**Reply by rarkins** [reply]
It's not possible soon

**Reply by ned-si** [reply]
Thanks for letting us know!
